hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java

```diff
@@ -410,61 +416,183 @@ public boolean rename(final Path src, final Path dst) throws IOException {
       }
-      return dstFileStatus.isDirectory() ? false : true;
+      return isDstDirectory.get() ? false : true;
     }
 
+    final AtomicBoolean isDstDirectory = new AtomicBoolean();
+    final AtomicBoolean isDstExists = new AtomicBoolean();
+
     // Non-HNS account need to check dst status on driver side.
-    if (!abfsStore.getIsNamespaceEnabled(tracingContext) && dstFileStatus == null) {
-      dstFileStatus = tryGetFileStatus(qualifiedDstPath, tracingContext);
+    if (!abfsStore.getIsNamespaceEnabled(tracingContext)) {
+      getPathInformation(qualifiedDstPath, tracingContext, isDstDirectory,
+          isDstExists);
     }
 
     try {
       String sourceFileName = src.getName();
       Path adjustedDst = dst;
 
-      if (dstFileStatus != null) {
-        if (!dstFileStatus.isDirectory()) {
+      if (isDstExists.get()) {
+        if (!isDstDirectory.get()) {
           return qualifiedSrcPath.equals(qualifiedDstPath);
         }
         adjustedDst = new Path(dst, sourceFileName);
+        qualifiedDstPath = makeQualified(adjustedDst);
+        LOG.debug("Qualified dst path: {}", qualifiedDstPath);
+        /*
+         * For blob-endpoint with nonHNS account, check if the qualifiedDstPath
+         * exist in backend. If yes, HTTP_CONFLICT exception has to be thrown.
+         */
+        if (getAbfsStore().getAbfsConfiguration().getPrefixMode()
+            == PrefixMode.BLOB) {
+          isDstDirectory.set(false);
+          isDstExists.set(false);
+          getPathInformation(qualifiedDstPath, tracingContext, isDstDirectory,
+              isDstExists);
+          if (isDstExists.get()) {
+            //destination already there. Rename should not be overwriting.
+            LOG.info("Rename src: {} dst: {} failed as qualifiedDst already exists",
+                qualifiedSrcPath, qualifiedDstPath);
+            throw new AbfsRestOperationException(
+                HttpURLConnection.HTTP_CONFLICT,
+                AzureServiceErrorCode.PATH_ALREADY_EXISTS.getErrorCode(), null,
+                null);
+          }
+        }
+      } else {
+        LOG.debug("dst {} doesn't exists. Check if the parent exists.", adjustedDst);
+        qualifiedDstPath = makeQualified(adjustedDst);
+        /*
+         * If the destination doesn't exist, check if parent of destination exists.
+         */
+        Path parent = qualifiedDstPath.getParent();
+        if (parent == null || !parent.isRoot()) {
+          isDstDirectory.set(false);
+          isDstExists.set(false);
+          getPathInformation(parent, tracingContext, isDstDirectory,
+              isDstExists);
+          if (!isDstExists.get() || !isDstDirectory.get()) {
+            LOG.info("parent of {} is {} doesn't exists. Failing rename",
+                adjustedDst, parent);
+            throw new AbfsRestOperationException(
+                HttpURLConnection.HTTP_NOT_FOUND,
+                RENAME_DESTINATION_PARENT_PATH_NOT_FOUND.getErrorCode(), null,
```

Comment:
Same as above, should exception message be null ?

Reply:
error code is same as DFS response.
error message doesn't matter, because it will be absorbed by https://github.com/saxenapranav/hadoop/blob/3af02586dfe553469dd040ca19b7a430336a2c3b/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java#L529-L539,
https://github.com/saxenapranav/hadoop/blob/3af02586dfe553469dd040ca19b7a430336a2c3b/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java#L1546-L1548

The rename method will give false and not throw exception